controllers/cluster_controller.go

```diff
@@ -124,6 +124,19 @@ func (r *ClusterReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr e
 		return ctrl.Result{}, err
 	}
 
+	defer func() {
```

Comment:
Uhmm, sorry to iterate on this point but It seems to me that by introducing the multiple defer statements that are executed in inverse order the code becomes less readable.
If this is required I think we should document why we are doing this and the expected behavior (update status before paching) before the second defer statement, but TBH If possible I would avoid this solution...
@vincepri opinions^^